src/main/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/segmentgroup/InvolvedParty.java

```diff
@@ -0,0 +1,69 @@
+package uk.nhs.digital.nhsconnect.lab.results.model.edifact.segmentgroup;
+
+import lombok.Getter;
+import lombok.NonNull;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.MessageRecipientNameAndAddress;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.PerformingOrganisationNameAndAddress;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.Reference;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.ReferenceType;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.RequesterNameAndAddress;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.ServiceProvider;
+
+import java.util.List;
+import java.util.Optional;
+
+import static java.util.stream.Collectors.toList;
+import static uk.nhs.digital.nhsconnect.lab.results.model.edifact.Segment.PLUS_SEPARATOR;
+
+/**
+ * Represents information about involved healthcare parties.
+ * <p>
+ * Segment group 1: {@code S01-NAD-ADR-COM-RFF-SEQ-SPR}
+ * <ul>
+ *     <li>{@code NAD} is optional. Must be qualified {@code +SLA}, {@code +PO}, or {@code +MR}, if present.</li>
+ *     <li>{@code ADR} is not used.</li>
+ *     <li>{@code COM} is not used.</li>
+ *     <li>{@code RFF} is optional. Must be qualified {@code +AHI}, if present.</li>
+ *     <li>{@code SEQ} is not used.</li>
+ *     <li>{@code SPR} is mandatory.</li>
+ * </ul>
+ * <p>
+ * Parents: {@link uk.nhs.digital.nhsconnect.lab.results.model.edifact.Message Message}
+ */
+@SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+public class InvolvedParty extends SegmentGroup {
+    public static final String INDICATOR = "S01";
+
+    @Getter(lazy = true)
+    private final Optional<PerformingOrganisationNameAndAddress> performingOrganisationNameAndAddress =
+        extractOptionalSegment(PerformingOrganisationNameAndAddress.KEY_QUALIFIER)
+            .map(PerformingOrganisationNameAndAddress::fromString);
+
+    @Getter(lazy = true)
+    private final Optional<RequesterNameAndAddress> requesterNameAndAddress =
+        extractOptionalSegment(RequesterNameAndAddress.KEY_QUALIFIER)
+            .map(RequesterNameAndAddress::fromString);
+
+    @Getter(lazy = true)
+    private final Optional<MessageRecipientNameAndAddress> messageRecipientNameAndAddress =
+        extractOptionalSegment(MessageRecipientNameAndAddress.KEY_QUALIFIER)
+            .map(MessageRecipientNameAndAddress::fromString);
```

Comment:
How would we feel about unifying the addresses too? They certainly do have different requirements, but perhaps all we care about is if we can get the info we need for translation to FHIR?

I'm inclined to say we should keep this in mind, but leave it for now until we actually try to produce the FHIR from them and see how we're using them.

Reply:
I agree - I think we should keep this in mind, but leave it for now until we get to the mapper stories. 

I've worked on mapping `PerformingOrganisationNameAndAddress` (which I've renamed to `PerformerNameAndAddress`, because it can be a Practitioner or a Organisation) in #51 and I think it might be possible, but I think we should wait until we try to produce FHIR from them.

Reply:
I agree with leaving this for now and keeping it in mind for another time 👍